<img src="https://qph.fs.quoracdn.net/main-qimg-757b494adc114bcef4b312661687f8b8.webp" width ="200">
<h1 style="text-align:center;"> <u>Course</u>: Foundation of Data Science(CS F320)</h1>
<h2 style="text-align:center;"><u>Instructor</u> : Dr Navneet Goel</h2>
 <h2 style="text-align:center;"> <u>Instructor</u> : Dr Vishal Gupta</h2>


<h1 style="text-align:center;"><u>Topic</u> : <span style="color:green">Money Ball Case Study and Analysis of data using Sabermetrics</span></h1>  
<img src="https://d13ezvd6yrslxm.cloudfront.net/wp/wp-content/images/moneyball-full-trailer.jpg" width="70%">

<h2>Background     :</h2>  
<h4>Source:Wikipedia<h4>

<h3>The 2002 Oakland A's</h3>

<p>The Oakland Athletics' 2002 season was the team's 35th in Oakland, California. It was also the 102nd season in franchise history. The Athletics finished first in the American League West with a record of 103-59.

The Athletics' 2002 campaign ranks among the most famous in franchise history. Following the 2001 season, Oakland saw the departure of three key players (the lost boys). Billy Beane, the team's general manager, responded with a series of under-the-radar free agent signings. The new-look Athletics, despite a comparative lack of star power, surprised the baseball world by besting the 2001 team's regular season record. The team is most famous, however, for winning 20 consecutive games between August 13 and September 4, 2002.[1] The Athletics' season was the subject of Michael Lewis' 2003 book Moneyball: The Art of Winning an Unfair Game (as Lewis was given the opportunity to follow the team around throughout that season)

This project is based off the book written by Michael Lewis (later turned into a movie).
</p>

<h3>Moneyball Book</h3>

<p>The central premise of book Moneyball is that the collective wisdom of baseball insiders (including players, managers, coaches, scouts, and the front office) over the past century is subjective and often flawed. Statistics such as stolen bases, runs batted in, and batting average, typically used to gauge players, are relics of a 19th-century view of the game and the statistics available at that time. The book argues that the Oakland A's' front office took advantage of more analytical gauges of player performance to field a team that could better compete against richer competitors in Major League Baseball (MLB).

Rigorous statistical analysis had demonstrated that on-base percentage and slugging percentage are better indicators of offensive success, and the A's became convinced that these qualities were cheaper to obtain on the open market than more historically valued qualities such as speed and contact. These observations often flew in the face of conventional baseball wisdom and the beliefs of many baseball scouts and executives.

By re-evaluating the strategies that produce wins on the field, the 2002 Athletics, with approximately US 44 million dollars in salary, were competitive with larger market teams such as the New York Yankees, who spent over US$125 million in payroll that same season.

</p>

<h3><i>1. What problem did the Oakland Athletics Baseball team (popularly known as A’s) faced?  </h3>
<p>
Oakland Athletics general manager Billy Beane wass hurt by his team's loss to the New York Yankees in the 2001 American League Division Series. With the impending departure of star players Johnny Damon, Jason Giambi, and Jason Isringhausen to free agency, Beane needed to assemble a competitive team for 2002 with Oakland's limited budget. 
  </p>

<h3><i>2. How did the General Manager, Billy Beane and Assistant General Manager, Paul DePodesta solve the problem?</i></h3>
<p>During a scouting visit to the Cleveland Indians, Beane met Paul DePodesta, a young Yale economics graduate with radical ideas about how to assess player value. Beane tested DePodesta's theory by asking whether he would have drafted Beane out of high school; though scouts considered Beane hugely promising, his career in the major leagues was disappointing.Podesta admited that, based on his method of assessing player value, he would not have drafted him until the ninth round. Impressed, Beane hired Brand as his assistant general manager. 
</p>

<h3><i>3.How the A’s identified players to be bought (Refer to Sabermetrics)? Build a suitable model. </i></h3>



<a href="Batting.csv"><strong><i>Batting</i><strong></a> 
    <br>
<a href="Salaries.csv"><strong><i>Salaries</i><strong></a> 

In [2]:
options(warn=-1)

library(dplyr)
library(ggplot2)



Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [3]:
batting <- read.csv("batting.csv")
salary <- read.csv("Salaries.csv")


In [4]:
options(repr.matrix.max.rows=nrow(batting), repr.matrix.max.cols=ncol(batting))
batting%>% arrange(yearID) %>% head()

playerID  yearID stint teamID lgID G  G_batting AB  R  H  X2B X3B HR RBI SB
1 abercda01 1871   1     TRO    NA    1  1          4  0  0  0  0   0   0  0 
2 addybo01  1871   1     RC1    NA   25 25        118 30 32  6  0   0  13  8 
3 allisar01 1871   1     CL1    NA   29 29        137 28 40  4  5   0  19  3 
4 allisdo01 1871   1     WS3    NA   27 27        133 28 44 10  2   2  27  1 
5 ansonca01 1871   1     RC1    NA   25 25        120 29 39 11  3   0  16  6 
6 armstbo01 1871   1     FW1    NA   12 12         49  9 11  2  1   0   5  0 
  CS BB SO IBB HBP SH SF GIDP G_old
1 0  0  0  NA  NA  NA NA NA    1   
2 1  4  0  NA  NA  NA NA NA   25   
3 1  2  5  NA  NA  NA NA NA   29   
4 1  0  2  NA  NA  NA NA NA   27   
5 2  2  1  NA  NA  NA NA NA   25   
6 1  0  1  NA  NA  NA NA NA   12

In [5]:
options(repr.matrix.max.rows=nrow(salary), repr.matrix.max.cols=ncol(salary))
salary %>% arrange(yearID) %>% head()

yearID teamID lgID playerID  salary 
1 1985   BAL    AL   murraed02 1472819
2 1985   BAL    AL   lynnfr01  1090000
3 1985   BAL    AL   ripkeca01  800000
4 1985   BAL    AL   lacyle01   725000
5 1985   BAL    AL   flanami01  641667
6 1985   BAL    AL   boddimi01  625000

## Analyzing the Lost Players

As previously mentioned, the Oakland A's lost 3 key players during the off-season. We'll want to get their stats to see what we have to replace. The players lost were: first baseman 2000 AL MVP Jason Giambi (giambja01) to the New York Yankees, outfielder Johnny Damon (damonjo01) to the Boston Red Sox and infielder Rainer Gustavo "Ray" Olmedo ('saenzol01').

Therefore we need players with good batting, infielding, and outfielding skills.


<h3><strong>Money available</strong> : <h5><i>$15 million</i></h5><h3>

<h2>Batting Table</h2>
<ul>
    <li>playerID : Player ID code</li>
    <li>yearID : Year</li>
    <li>stint : player's stint (order of appearances within a season)</li>
    <li>teamID : Team</li>
    <li>lgID : League</li>
    <li>G : Games</li>
    <li>AB : At Bats</li>
    <li>R : Runs</li>
    <li>H : Hits</li>
    <li>2B : Doubles</li>
    <li>3B : Triples</li>
    <li>HR : Homeruns</li>
    <li>RBI : Runs Batted In</li>
    <li>SB : Stolen Bases</li>
    <li>CS : Caught Stealing</li>
    <li>BB : Base on Balls</li>
    <li>SO : Strikeouts</li>
    <li>IBB : Intentional walks</li>
    <li>HBP : Hit by pitch</li>
    <li>SH : Sacrifice hits</li>
    <li>SF : Sacrifice flies</li>
    <li>GIDP : Grounded into double plays</li>
</ul>    

<h2>Fielding Table</h2>
<ul>
<li>    yearID     :    Year</li>
<li>    stint      :    player's stint (order of appearances within a season)</li>
<li>    teamID     :    Team</li>
<li>    lgID       :    League</li>
<li>    Pos        :    Position</li>
<li>    G          :    Games </li>
<li>    GS         :    Games Started</li>
<li>    InnOuts    :    Time played in the field expressed as outs </li>
<li>    PO         :    Putouts</li>
<li>    A          :    Assists</li>
<li>    E          :    Errors</li>
<li>    DP         :    Double Plays</li>
<li>    PB         :    Passed Balls (by catchers)</li>
<li>    WP         :    Wild Pitches (by catchers)</li>
<li>    SB         :    Opponent Stolen Bases (by catchers)</li>
<li>    CS         :    Opponents Caught Stealing (by catchers)</li>
<li>    ZR         :    Zone Rating</li>
<li>    playerID   :    Player ID code</li>
</ul>


<h2>Feature Engineering:    <i>Batting</i></h2>
<ul>
   <li> Batting Average :<br>
        \begin{align}    
        BA = \frac{H}{AB} 
        \end{align}
   </li><br> 
   <li> On Base Percentage :
       <br> <br>
        \begin{align}    
        OBP = \frac{H + BB + HBP}{AB + BB + HBP +SF } 
        \end{align}
   </li><br>    
   <li> Slugging Percentage : 
       <br><br>
       \begin{align}    
        SP = \frac{1 B + (2 \times 2B) + (3 \times 3B) +(4 \times HR)}{AB} 
        \end{align}
   </li>    
   <br><br>
    <li> Runs per game : 
       <br><br>
       \begin{align}    
        AR = \frac{R}{G} 
        \end{align}
   </li>    
   <br><br>
   \begin{align}    
    H = \ B + 2B + 3B + HR  
   \end{align}


   


In [6]:
batting <- na.omit(batting, cols=c(AB:HR,"HBP","SF"))
batting <- batting %>% mutate(BA=H/AB, OBP=(H+BB+HBP)/(AB+BB+HBP+SF), B=H-(X2B+X3B+HR), SP=(B+2*X2B+3*X3B+4*HR)/AB, score=(1+.7*.7)*(OBP*SP)/(0.7*.7*OBP+SP), AR=R/G)


In [7]:
batting <- batting %>% filter(!is.nan(score), !is.nan(AR)) 


In [8]:
options(repr.matrix.max.rows=nrow(batting), repr.matrix.max.cols=ncol(batting))

head(batting)

playerID  yearID stint teamID lgID G   G_batting AB  R   H   X2B X3B HR RBI
1 aaronha01 1955   1     ML1    NL   153 153       602 105 189 37   9  27 106
2 aaronha01 1956   1     ML1    NL   153 153       609 106 200 34  14  26  92
3 aaronha01 1957   1     ML1    NL   151 151       615 118 198 27   6  44 132
4 aaronha01 1958   1     ML1    NL   153 153       601 109 196 34   4  30  95
5 aaronha01 1959   1     ML1    NL   154 154       629 116 223 46   7  39 123
6 aaronha01 1960   1     ML1    NL   153 153       590 102 172 20  11  40 126
  SB CS BB SO IBB HBP SH SF GIDP G_old BA        OBP       B   SP       
1  3 1  49 61  5  3   7   4 20   153   0.3139535 0.3662614 116 0.5398671
2  2 4  37 54  6  2   5   7 21   153   0.3284072 0.3648855 126 0.5582923
3  1 1  57 58 15  0   0   3 13   151   0.3219512 0.3777778 121 0.6000000
4  4 1  59 49 16  1   0   3 21   153   0.3261231 0.3855422 128 0.5457571
5  8 0  51 54 17  4   0   9 19   154   0.3545310 0.4011544 131 0.6359300
6 16 7  60 63 13  2   0  12  8   153   0.2915254 0.3524096 101 0.5661017
  score     AR       
1 0.4095746 0.6862745
2 0.4117999 0.6928105
3 0.4301727 0.7814570
4 0.4267402 0.7124183
5 0.4565887 0.7532468
6 0.4023574 0.6666667

In [9]:
batting19 <- batting %>% filter((1985<=yearID) )
batting19 <- merge(batting, salary)
batting20 <- batting19 %>% filter(yearID >2001)
batting19 <- batting19 %>% filter(yearID <=2001)

In [10]:
batting19 <- batting19 %>% arrange(yearID)  #data 1885 to 2001 ()
batting20 <- batting20 %>% arrange(yearID)  #data 2001 and beyond
head(batting19)

playerID  yearID teamID lgID stint G   G_batting AB  R  H   X2B X3B HR RBI SB
1 aguaylu01 1985   PHI    NL   1      91  91       165 27  46  7  3    6 21   1
2 almonbi01 1985   PIT    NL   1      88  88       244 33  66 17  0    6 29  10
3 andujjo01 1985   SLN    NL   1      38  38        94  2  10  2  0    0  8   3
4 armasto01 1985   BOS    AL   1     103 103       385 50 102 17  5   23 64   0
5 ashbyal01 1985   HOU    NL   1      65  65       189 20  53  8  0    8 25   0
6 ayalabe01 1985   CLE    AL   1      46  46        76 10  19  7  0    2 15   0
  ... BB SO IBB HBP SH SF GIDP G_old BA        OBP       B  SP        score    
1 ... 22 26 5   6   4  3   7    91   0.2787879 0.3775510 30 0.4666667 0.4028498
2 ... 22 61 0   1   4  3   6    88   0.2704918 0.3296296 43 0.4139344 0.3532924
3 ...  5 50 0   0   7  0   2    38   0.1063830 0.1515152  8 0.1276596 0.1427431
4 ... 18 90 4   2   0  5  14   103   0.2649351 0.2975610 57 0.5142857 0.3454325
5 ... 24 27 2   1   1  1   9    65   0.2804233 0.3627907 37 0.4497354 0.3874216
6 ...  4 17 1   0   0  1   2    46   0.2500000 0.2839506 10 0.4210526 0.3180031
  AR         salary 
1 0.29670330  237000
2 0.37500000  255000
3 0.05263158 1030000
4 0.48543689  915000
5 0.30769231  416667
6 0.21739130  303333

In [11]:
batters_pool <- batting19 %>% filter(ifelse(yearID==2001,teamID!="OAK",T) ) #select players who are currently not in team


In [12]:
team <-batting19 %>% filter(teamID=="OAK", yearID==2001 )
runs_total <- sum(team$R)

In [13]:
options(repr.matrix.max.rows=nrow(batting19), repr.matrix.max.cols=ncol(batting19)) ##view all cols

lost_players <- team %>% filter(playerID %in% c("giambja01",'damonjo01','saenzol01')) 
lost_players

playerID  yearID teamID lgID stint G   G_batting AB  R   H   X2B X3B HR RBI
1 damonjo01 2001   OAK    AL   1     155 155       644 108 165 34  4    9  49
2 giambja01 2001   OAK    AL   1     154 154       520 109 178 47  2   38 120
3 saenzol01 2001   OAK    AL   1     106 106       305  33  67 21  1    9  32
  SB CS BB  SO IBB HBP SH SF GIDP G_old BA        OBP       B   SP       
1 27 12  61 70  1   5  5  4   7   155   0.2562112 0.3235294 118 0.3633540
2  2  0 129 83 24  13  0  9  17   154   0.3423077 0.4769001  91 0.6596154
3  0  1  19 64  1  13  1  3   9   106   0.2196721 0.2911765  36 0.3836066
  score     AR        salary 
1 0.3356267 0.6967742 7100000
2 0.5246974 0.7077922 4103333
3 0.3162345 0.3113208  290000

In [14]:
team_without <- team %>% filter(!(playerID %in% c("giambja01",'damonjo01','saenzol01')) )

In [15]:
runs_partial <- sum(team_without$R)
team_without

playerID  yearID teamID lgID stint G   G_batting AB  R   H   X2B X3B HR RBI
1  chaveer01 2001   OAK    AL   1     151 151       552  91 159 43  0   32 114
2  fasansa01 2001   OAK    AL   1      11  11        21   2   1  0  0    0   0
3  giambje01 2001   OAK    AL   1     124 124       371  64 105 26  0   12  57
4  heredgi01 2001   OAK    AL   1      24   1         3   0   1  0  0    0   0
5  hernara02 2001   OAK    AL   1     136 136       453  55 115 25  0   15  60
6  jahajo01  2001   OAK    AL   1      12  12        45   2   4  3  0    0   8
7  jenniro01 2001   OAK    AL   1      20  20        52   4  13  3  0    0   4
8  longte01  2001   OAK    AL   1     162 162       629  90 178 37  4   12  85
9  menecfr01 2001   OAK    AL   1     139 139       471  82 114 22  2   12  60
10 muldema01 2001   OAK    AL   1      34   2         5   0   1  0  0    0   0
11 ortizjo02 2001   OAK    AL   1      11  11        42   4   7  0  0    0   3
12 piattad01 2001   OAK    AL   1      36  36        95   9  20  5  1    0   6
13 tejadmi01 2001   OAK    AL   1     162 162       622 107 166 31  3   31 113
14 valdema02 2001   OAK    AL   1      32  32        54   7  15  1  0    1   8
   SB CS BB SO  IBB HBP SH SF GIDP G_old BA         OBP       B   SP        
1   8 2  41  99 9    4  0  7   7   151   0.28804348 0.3377483  84 0.53985507
2   0 0   1  12 0    1  0  0   1    11   0.04761905 0.1304348   1 0.04761905
3   0 1  63  83 1    4  3  2  13   124   0.28301887 0.3909091  67 0.45013477
4   0 0   0   0 0    0  0  0   0     1   0.33333333 0.3333333   1 0.33333333
5   1 1  37  68 3    6  9  4  10   136   0.25386313 0.3160000  75 0.40838852
6   0 0   6  15 0    0  0  1   3    12   0.08888889 0.1923077   1 0.15555556
7   0 0   2   6 0    0  0  1   0    20   0.25000000 0.2727273  10 0.30769231
8   9 3  52 103 8    0  0  6  17   162   0.28298887 0.3347889 125 0.41176471
9   2 3  79  97 0   19  3  6  13   139   0.24203822 0.3686957  78 0.37367304
10  0 0   0   1 0    0  0  0   0     2   0.20000000 0.2000000   1 0.20000000
11  1 0   3   5 0    0  0  1   4    11   0.16666667 0.2173913   7 0.16666667
12  0 0  13  26 0    0  1  2   5    36   0.21052632 0.3000000  14 0.28421053
13 11 5  43  89 5   13  1  4  14   162   0.26688103 0.3255132 101 0.47588424
14  0 0  12  18 1    1  0  0   0    32   0.27777778 0.4179104  13 0.35185185
   score      AR        salary 
1  0.38516865 0.6026490  625000
2  0.08297754 0.1818182  375000
3  0.40858832 0.5161290  282500
4  0.33333333 0.0000000 3300000
5  0.34139900 0.4044118  245000
6  0.17844311 0.1666667 3250000
7  0.28331485 0.2000000  290000
8  0.35671906 0.5555556  240750
9  0.37031781 0.5899281  225000
10 0.20000000 0.0000000  232500
11 0.19761273 0.3636364  201000
12 0.29461736 0.2500000  217500
13 0.36326098 0.6604938 2025000
14 0.39360836 0.2187500  205000

In [16]:
runs_partial

[1] 517

In [17]:
difference <- runs_total - runs_partial
difference

[1] 250

In [18]:
money_we_have <- 15000000-sum(team_without$salary)
money_we_have

[1] 3285750

In [19]:
lost_players <- lost_players %>% arrange(OBP,SP)


In [20]:
lost_players %>% select(playerID,OBP,SP,G,R) 

playerID  OBP       SP        G   R  
1 saenzol01 0.2911765 0.3836066 106  33
2 damonjo01 0.3235294 0.3633540 155 108
3 giambja01 0.4769001 0.6596154 154 109

<link href="https://fonts.googleapis.com/css?family=Dokdo" rel="stylesheet">
<u><h1>How to choose players ?- <span style="color:yellow;font-style:italic;font-family: 'Dokdo', cursive;">Algorithm</span></h1></u>

<p><i>We iterate through the rows picking 3 rows at a time and then multiplying number of games offered to the three lost players to corresponding players in batters_pool(the three rows) by multiplying max games to player with highest OBP,and then SP since OBP determines the ability of the player to remain at base and stay longer in game.  </i></p>

In [21]:
players_to_choose <- data.frame(Player1=" ", Player2=" ", Player3=" ",R=0,Total_spendings=0 ,stringsAsFactors = F) 
i = 1

for (row1 in 1:nrow(tail(batters_pool,50))){
 for (row2 in row1:nrow(tail(batters_pool,50)))
    for (row3 in row2:nrow(tail(batters_pool,20))){
      temp <- rbind(batters_pool[row1,],batters_pool[row2,],batters_pool[row3,]) 
      temp<- temp %>% arrange(OBP, SP) 
      for (k in 1:3){
        if (temp[k,"AR"] >= lost_players[k,"AR"]){
           temp[k,"R"] <- temp[k,"AR"]*lost_players[k,"G"]}  }
      if ((batters_pool[row1,]$salary + batters_pool[row2,]$salary + batters_pool[row3,]$salary <= money_we_have)&
        (batters_pool[row1,]$R + batters_pool[row2,]$R + batters_pool[row3,]$R >=250))          {
         players_to_choose<- rbind(players_to_choose,c(as.character(batters_pool[row1,]$playerID) ,as.character(batters_pool[row2,]$playerID), as.character(batters_pool[row3,]$playerID), batters_pool[row1,]$R + batters_pool[row2,]$R + batters_pool[row3,]$R, batters_pool[row1,]$salary + batters_pool[row2,]$salary + batters_pool[row3,]$salary)) 
         i <- i + 1       
}}}

<link href="https://fonts.googleapis.com/css?family=Dokdo" rel="stylesheet">
<u><h1><span style="color:#00BFFF;font-style:italic;font-family: 'Dokdo', cursive;">Few triples of players which can be selected </span></h1></u>


In [22]:
players_to_choose %>%filter (Player1!=Player2,Player2!=Player3)

Player1   Player2   Player3   R   Total_spendings
1   armasto01 bradlph01 boggswa01 257 2040000        
2   armasto01 brettge01 bradlph01 258 2040000        
3   armasto01 brettge01 boggswa01 265 2915000        
4   backmwa01 baineha01 barfije01 257 1200000        
5   backmwa01 baineha01 bellge02  250 1210000        
6   backmwa01 barfije01 bellge02  258 860000         
7   backmwa01 boggswa01 bernato01 257 1600000        
8   backmwa01 boggswa01 bellge02  271 1535000        
9   backmwa01 bradlph01 boggswa01 284 1325000        
10  backmwa01 bradlph01 bernato01 250 725000         
11  backmwa01 bradlph01 bellge02  264 660000         
12  backmwa01 brettge01 bradlph01 285 1325000        
13  backmwa01 brettge01 boggswa01 292 2200000        
14  backmwa01 brettge01 bernato01 258 1600000        
15  backmwa01 brettge01 bellge02  272 1535000        
16  backmwa01 brookhu01 brettge01 252 1750000        
17  backmwa01 brookhu01 boggswa01 251 1750000        
18  backmwa01 brunato01 brettge01 256 1625000        
19  backmwa01 brunato01 boggswa01 255 1625000        
20  backmwa01 bucknbi01 brettge01 274 1947500        
21  backmwa01 bucknbi01 bradlph01 266 1072500        
22  backmwa01 bucknbi01 boggswa01 273 1947500        
23  backmwa01 bucknbi01 bellge02  253 1282500        
24  baineha01 balbost01 barfije01 254 1205000        
25  baineha01 barfije01 basske01  252 1155000        
26  baineha01 barfije01 baylodo01 250 1810000        
27  baineha01 barfije01 bellge02  267 1335000        
28  baineha01 boggswa01 bernato01 266 2075000        
29  baineha01 boggswa01 bellge02  280 2010000        
30  baineha01 bradlph01 boggswa01 293 1800000        
31  baineha01 bradlph01 bernato01 259 1200000        
32  baineha01 bradlph01 bellge02  273 1135000        
33  baineha01 brettge01 bradlph01 294 1800000        
34  baineha01 brettge01 boggswa01 301 2675000        
35  baineha01 brettge01 bernato01 267 2075000        
36  baineha01 brettge01 bellge02  281 2010000        
37  baineha01 brockgr01 brettge01 258 1825000        
38  baineha01 brockgr01 bradlph01 250 950000         
39  baineha01 brockgr01 boggswa01 257 1825000        
40  baineha01 brookhu01 brettge01 261 2225000        
41  baineha01 brookhu01 bradlph01 253 1350000        
42  baineha01 brookhu01 boggswa01 260 2225000        
43  baineha01 brunato01 brettge01 265 2100000        
44  baineha01 brunato01 bradlph01 257 1225000        
45  baineha01 brunato01 boggswa01 264 2100000        
46  baineha01 bucknbi01 brettge01 283 2422500        
47  baineha01 bucknbi01 bradlph01 275 1547500        
48  baineha01 bucknbi01 boggswa01 282 2422500        
49  baineha01 bucknbi01 bellge02  262 1757500        
50  bakerdu01 bradlph01 boggswa01 255 1700000        
51  bakerdu01 brettge01 bradlph01 256 1700000        
52  bakerdu01 brettge01 boggswa01 263 2575000        
53  balbost01 barfije01 bellge02  255 865000         
54  balbost01 boggswa01 bernato01 254 1605000        
55  balbost01 boggswa01 bellge02  268 1540000        
56  balbost01 bradlph01 boggswa01 281 1330000        
57  balbost01 bradlph01 bellge02  261 665000         
58  balbost01 brettge01 bradlph01 282 1330000        
59  balbost01 brettge01 boggswa01 289 2205000        
60  balbost01 brettge01 bernato01 255 1605000        
61  balbost01 brettge01 bellge02  269 1540000        
62  balbost01 brunato01 brettge01 253 1630000        
63  balbost01 brunato01 boggswa01 252 1630000        
64  balbost01 bucknbi01 brettge01 271 1952500        
65  balbost01 bucknbi01 bradlph01 263 1077500        
66  balbost01 bucknbi01 boggswa01 270 1952500        
67  balbost01 bucknbi01 bellge02  250 1287500        
68  barfije01 basske01  bellge02  253 815000         
69  barfije01 baylodo01 bellge02  251 1470000        
70  barfije01 bernato01 bellge02  254 1060000        
71  barfije01 boggswa01 bernato01 274 1725000        
72  barfije01 boggswa01 beniqju01 255 1690000        
73  barfije01 boggswa01 bellge02  288 1660000        
74  barf

In [23]:
batters_pool <- rbind(batters_pool,team)


<link href="https://fonts.googleapis.com/css?family=Dokdo" rel="stylesheet">
<h1><u>Model:</u><span style="color:green;font-style:italic;font-family: 'Dokdo', cursive;"> Logistic Regression</span></h1>

<p>Predicting whether a team would join class 0 or not joining OAK's or 1  join OAK's. For this I take the data from players pool and assign 1 if he belongs to "OAK" and 0 otherwise </p>

In [24]:
batters_2002 <- batting20 %>%filter(teamID =="OAK", yearID==2002) #players who are in batters list of 2002 and in OAK
batters_2002 <- batters_2002 %>% filter(!(playerID %in% team_without$playerID)) 
test <- batting20 %>% filter(yearID==2002) %>% mutate(label=ifelse(playerID %in% batters_2002$playerID,1,0)) #test data

In [25]:
train <- batting19 %>% mutate(label=ifelse(teamID=="OAK",1,0)) # assign class 1 to in OAK and 0 to not in OAK
logistic <- glm( label~ OBP + SP + AR  + BA , data = train,family = binomial)

In [26]:
summary(logistic) 


Call:
glm(formula = label ~ OBP + SP + AR + BA, family = binomial, 
    data = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.6011  -0.2718  -0.2362  -0.2014   2.9534  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept) -4.34565    0.23673 -18.357  < 2e-16 ***
OBP          3.55284    1.12874   3.148  0.00165 ** 
SP          -0.02209    0.79948  -0.028  0.97796    
AR           1.61287    0.35352   4.562 5.06e-06 ***
BA          -3.31297    1.50797  -2.197  0.02802 *  
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2540.3  on 9403  degrees of freedom
Residual deviance: 2494.4  on 9399  degrees of freedom
AIC: 2504.4

Number of Fisher Scoring iterations: 6


In [27]:
li.null <-logistic$null.deviance/-2
li.proposed <-logistic$deviance/-2
(li.null-li.proposed)/li.null

[1] 0.01808543

#### By above result  p value <.1 and thus null hypothesis is quite significant.Too bad for the model...

In [29]:
x <- predict(logistic, test, type="response") 

In [30]:
dt <- test %>% mutate(predicted = ifelse(x>  0.5,1, 0))   #alpha- value here gives high misclassification
x_ <- dt %>% filter(label != predicted) %>% count()

In [31]:
x_ <- dt %>% filter(label != predicted) %>% count()
x_  #Incorrect Classification

n
1 9

In [32]:
x_ <- dt %>% filter(label == predicted) %>% count()
x_  #Correct Classification

n  
1 562


<link href="https://fonts.googleapis.com/css?family=Dokdo|ZCOOL+XiaoWei" rel="stylesheet">

<h1>
    <span style="font-family: 'Dokdo', cursive;color:red;font-style:italic;border: 2px solid red;
  border-radius: 5px;">F1_Score</span></h1> 

In [33]:
TN <- dt %>% filter(label == predicted, label==0) %>% count()%>% as.numeric()
TP <- dt %>% filter(label == predicted, label==1) %>% count()%>% as.numeric()
FP <- dt %>% filter(label != predicted, label==1) %>% count()%>% as.numeric()
FN <- dt %>% filter(label != predicted, label==0) %>% count()%>% as.numeric()
TN
TP
FP
FN


[1] 562

[1] 0

[1] 9

[1] 0

In [34]:
precision <- TP/(TP+FN)
recall <- TP/(TP+TN)
precision 
recall

[1] NaN

[1] 0

In [35]:
F1_score <- 2*precision*recall/(precision+recall)
F1_score

[1] NaN

 \begin{align}    
  Accuracy = \frac{562}{562+9} 
       =  98.4\%
 \end{align}
 
 \begin{align}    
  F_{1} \thinspace Score = \frac{2*precision*recall}{precision+recall} 
           =  ???
  \end{align}
 
 \begin{align}   
  For \thinspace good \thinspace model \thinspace F_{1} \thinspace score \approx 1
 \end{align}

In [36]:
classified_correctly_by_model <- dt %>% filter(label==predicted,teamID=="OAK")
players_to_choose_1 <- players_to_choose[,1:3] 


<link href="https://fonts.googleapis.com/css?family=Dokdo|Gloria+Hallelujah|Pacifico|ZCOOL+XiaoWei" rel="stylesheet"> <h2><i><span style="color:#800000; font-family:'Pacifico', cursive;">  To check if players will join I would check if correctly classified points i.e where label is equal to predicted exists in classified_correctly_by_model$playerID</span><i></h2>

In [37]:
will_join <-c()
for (row in players_to_choose_1){
    temp <- as.character(row)
    count <-0
    vec <- c()
    for (j in temp)
        {if (j %in% classified_correctly_by_model$playerID)
            count <- count + 1
            vec<-append(vec,j)
        }
if (count == 3)
    print(paste("All 3: ",vec))
else
    print(paste("A few : ",vec))
 
}
    
    

  [1] "A few :   "         "A few :  armasto01" "A few :  armasto01"
  [4] "A few :  armasto01" "A few :  armasto01" "A few :  armasto01"
  [7] "A few :  armasto01" "A few :  backmwa01" "A few :  backmwa01"
 [10] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [13] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [16] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [19] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [22] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [25] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [28] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [31] "A few :  backmwa01" "A few :  backmwa01" "A few :  backmwa01"
 [34] "A few :  baineha01" "A few :  baineha01" "A few :  baineha01"
 [37] "A few :  baineha01" "A few :  baineha01" "A few :  baineha01"
 [40] "A few :  baineha01" "A few :  baineha01" "A few :  baineha01"
 [43] "A few :  baineha01" "A few 